src/js/reducers/conversation-reducer.js

```diff
@@ -36,22 +34,80 @@ const addMessage = (messages, message) => {
     return sortMessages([...messages, message]);
 };
 
+
+const matchMessage = (message, queryProps) => Object.keys(queryProps).every((key) => message[key] === queryProps[key]);
+
+const replaceMessage = (messages, query, newMessage) => {
+    const existingMessage = messages.find((message) => matchMessage(message, query));
+    if (!existingMessage) {
+        return messages;
+    }
+
+    const index = messages.indexOf(existingMessage);
+
+    // get rid of duplicate values
+    return [...messages.slice(0, index), newMessage, ...messages.slice(index + 1)].reduce((filteredMessages, nextMessage) => {
+        const message = filteredMessages.find((m) => isEqual(m, nextMessage));
+        if (message) {
+            return filteredMessages;
+        }
+        return [...filteredMessages, nextMessage];
+    }, []);
+};
+
+const isEqual = (a, b) => {
+    if (a._id && b._id && a._id === b._id) {
+        return true;
+    }
+
+    if (!a._id || !b._id) {
+        if (a.role === b.role) {
+            if (a.text && b.text && a.text === b.text) {
+                return true;
+            }
+
+            if (a.mediaType === b.mediaType && a.mediaUrl === b.mediaUrl) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+};
+
+const mergeMessages = (a, b) => {
+    // concat will make a union out of both arrays
+    // reduce will return a new array, the function used in the reduction strips out duplicates
+    return a.concat(b).reduce((filteredMessages, nextMessage) => {
```

Comment:
Hmm, this is only ever used in the context of the `reduce` though. Might be confusing if it's moved outside? @lemieux